presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/JoinPrefilter.java

```diff
@@ -34,18 +39,52 @@
 import java.util.List;
 import java.util.Map;
 import java.util.Optional;
+import java.util.stream.IntStream;
 
 import static com.facebook.presto.SystemSessionProperties.isJoinPrefilterEnabled;
+import static com.facebook.presto.common.type.BigintType.BIGINT;
 import static com.facebook.presto.common.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.common.type.VarcharType.VARCHAR;
 import static com.facebook.presto.spi.plan.AggregationNode.singleGroupingSet;
 import static com.facebook.presto.spi.plan.JoinType.INNER;
 import static com.facebook.presto.spi.plan.JoinType.LEFT;
+import static com.facebook.presto.sql.planner.PlannerUtils.addProjections;
 import static com.facebook.presto.sql.planner.PlannerUtils.clonePlanNode;
 import static com.facebook.presto.sql.planner.PlannerUtils.isScanFilterProject;
+import static com.facebook.presto.sql.planner.PlannerUtils.orNullHashCode;
 import static com.facebook.presto.sql.planner.PlannerUtils.projectExpressions;
+import static com.facebook.presto.sql.planner.PlannerUtils.restrictOutput;
 import static com.facebook.presto.sql.planner.plan.ChildReplacer.replaceChildren;
+import static com.facebook.presto.sql.relational.Expressions.call;
+import static com.facebook.presto.sql.relational.Expressions.callOperator;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
 import static java.util.Objects.requireNonNull;
 
+/**
+ * This optimizer filter the right side of a join with the unique join keys on the left side of the join. When the join key is wide or
+ * there are multiple join keys, we are to do filter on the hash instead of using the keys.
+ * It will convert plan from
+ * <pre>
+ *     - InnerJoin
+ *          leftKey = rightKey
+ *          - scan l
+ *          - scan r
+ * </pre>
+ * into
+ * <pre>
+ *     - InnerJoin
+ *          leftKey = rightKey
+ *          - scan l
+ *          - semijoin
```

Comment:
nit: Can show the hash project here to explain

Reply:
Updated to cover all three cases
* one join key (not varchar type)
* one join key (varchar type)
* multiple join keys